PathUtils: *always* prefer longer to shorter extension match

djudd opened the pull request:
The older logic would *usually* prefer a longer to a shorter extension
match, but not in the case with a three-or-more-component extension
when one of the intermediate extensions wasn't recognized, e.g.
".coffee" would be preferred over ".js.jsx.coffee" unless ".jsx.coffee"
was also a recognized extension.

(".js.jsx.coffee" is the preferred Coffee+React extension with
the react-rails gem)

djudd commented:
FWIW, my version is also about 2.5x faster on Ruby 2.2, according to my dumb microbenchmarks:

Old:
```bash
irb(main):004:0> puts Benchmark.measure { 100000.times { Sprockets::PathUtils.match_path_extname("foo/component.js.jsx", { '.js' => 'application/javascript', '.js.jsx' => 'application/jsx' }) } }
  0.820000   0.020000   0.840000 (  0.841651)
irb(main):002:0> puts Benchmark.measure { 100000.times { Sprockets::PathUtils.match_path_extname("foo/component.js", { '.js' => 'application/javascript', '.js.jsx' => 'application/jsx' }) } }
  0.750000   0.020000   0.770000 (  0.766332)
```

New:
```bash
irb(main):023:0> puts Benchmark.measure { 100000.times { Sprockets::PathUtils.match_path_extname("foo/component.js.jsx", { '.js' => 'application/javascript', '.js.jsx' => 'application/jsx' }) } }
  0.310000   0.000000   0.310000 (  0.311528)
irb(main):027:0> puts Benchmark.measure { 100000.times { Sprockets::PathUtils.match_path_extname("foo/component.js", { '.js' => 'application/javascript', '.js.jsx' => 'application/jsx' }) } }
  0.300000   0.010000   0.310000 (  0.307181)
```

schneems commented:
Great job. I did a bunch of benchmarks and it looks like this is much much faster. I've not been able to find a slower case. I think that String#scan is not very fast, also it allocates an array and looping through the array with `do` syntax creates a lexical scope that we avoid with the while loop. 

How did you find this piece and come up with this optimization?

schneems commented:
Backported to 3.x

djudd commented:
Thanks! I noticed the bug that I fixed here when I was trying to port a project that uses both `react-rails` and Coffeescript to Sprockets 4. I added a transformer for `.js.jsx.coffee` files, but it wasn't used, because `.coffee` was a recognized extension but `.jsx.coffee` was not, so it stopped at `.coffee`. Took me quite a while to figure out what was happening...

schneems commented:
Awesome. Did you plan on it being faster too or was that a side effect? Thanks for contributing to Sprockets btw! I released your backported patch in 3.6.3

djudd commented:
Since it looked like rewriting from scratch was needed, I did try to do so in a way that minimized allocations. Glad it worked out.
